****
# Text Preparation

### Before the start


The Penguin Classic version of David Hawkes and John Minford's translation is downloaded from the internet in a pdf. file. Then, I used ZAMZAR, an online file convertor, to convert the pdf. file into one single txt. file.

The full text is manually broken down into 6 parts: chapter 1-20, chapter 21-40, chapter 41-60, chapter 61-80, chapter 81-100 and chapter 101-120. These text files are named StoneA, StoneB, etc..
I do not follow Chan's way of breaking up the text, which divide the 120 chapters into three parts, because spaCy has a limitation on the length of text it can process at once.

Additionally, the book title and translator's information are manually removed.

### Import the text

In [34]:
## import pandas and print the working directory

import pandas as pd
%pwd

'/Users/chaguan/Desktop/DIGHUM101-2021-main/Final Project'

I import the three-part text and name each StoneA (includes chapter 1-40), StoneB (includes 41-80) and StoneC (includes 81-120) respectively. Here, only StoneA is printed for the sake of space.

In [2]:
StoneA = open("The story of the stone(1-20).txt").read()

In [48]:
print(StoneA)

CHAPTER 1



Zhen Shi-yin makes the Stone’s
acquaintance in a dream
And Jia Yu-cun finds that poverty is not incompatible
with romantic feelings

GENTLE READER,



What, you may ask, was the origin of this book?
Though the answer to this question may at first seem to border on the absurd,
reflection will show that there is a good deal more in it than meets the eye.
     Long ago, when the goddess Nǚ-wa was repairing the sky, she melted down a
great quantity of rock and, on the Incredible Crags of the Great Fable Mountains,
moulded the amalgam into thirty-six thousand, five hundred and one large building
blocks, each measuring seventy-two feet by a hundred and forty-four feet square.
She used thirty-six thousand five hundred of these blocks in the course of her
building operations, leaving a single odd block unused, which lay, all on its own, at
the foot of Greensickness Peak in the aforementioned mountains.
    Now this block of stone, having undergone the melting and molding of a godd

### Count words

The following information is checked in advance to give a basic picture of the text.

In [38]:
## How many words in the text?
print(len(StoneA.split()))

128567


In [39]:
## How many lines?
print(len(StoneA.split("\n")))

10974


In [4]:
## import the module for punctuation removal
from string import punctuation

In [6]:
## The following punctuation will be removed from the text:
print(punctuation)

!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~


In [9]:
## Remove the punctuation from the above list
for char in punctuation:
    StoneA = StoneA.replace(char, "")

In [10]:
print(StoneA)

CHAPTER 1



Zhen Shi yin makes the Stone’s
acquaintance in a dream
And Jia Yu cun finds that poverty is not incompatible
with romantic feelings

GENTLE READER 



What  you may ask  was the origin of this book 
Though the answer to this question may at first seem to border on the absurd 
reflection will show that there is a good deal more in it than meets the eye 
     Long ago  when the goddess Nǚ wa was repairing the sky  she melted down a
great quantity of rock and  on the Incredible Crags of the Great Fable Mountains 
moulded the amalgam into thirty six thousand  five hundred and one large building
blocks  each measuring seventy two feet by a hundred and forty four feet square 
She used thirty six thousand five hundred of these blocks in the course of her
building operations  leaving a single odd block unused  which lay  all on its own  at
the foot of Greensickness Peak in the aforementioned mountains 
    Now this block of stone  having undergone the melting and molding of a godd

#### Why I do not remove stopwords and lemmatize?

### Parse the Text Using spaCy

In [21]:
import spacy
nlp = spacy.load('en_core_web_lg')

In [46]:
## The following codes refer an answer on Stack, See: 
##https://stackoverflow.com/questions/51685224/percentage-count-verb-noun-using-spacy

from __future__ import unicode_literals
import spacy,en_core_web_sm
from collections import Counter
nlp = en_core_web_sm.load()
c = Counter(([token.pos_ for token in nlp(StoneA)]))
sbase = sum(c.values())
for el, cnt in c.items():
    print(el, '{0:2.2f}%'.format((100.0* cnt)/sbase))

NOUN 14.25%
NUM 0.75%
SPACE 7.18%
PROPN 5.59%
VERB 13.87%
DET 9.15%
PART 3.30%
ADP 10.87%
CCONJ 3.46%
AUX 4.81%
ADJ 5.28%
PRON 10.54%
SCONJ 1.47%
ADV 5.79%
INTJ 0.18%
PUNCT 3.49%
SYM 0.00%
X 0.01%


In [ ]:
## Notes for understanding the above chart:

## Abbriviation system (quoted from Stack Overflow
##((https://stackoverflow.com/questions/40288323/what-do-spacys-part-of-speech-and-dependency-tags-mean)): 
## NOUN=Noun; NUM=numeral; PROPN=proper noun; DET=determiner, like "a""an""the"; 
## (continued):PART=particle, like "'s""not"; ADP=adposition, like "in""to""during";
## (continued):CCONJ=coordinating conjunction, like "and""or""but"; AUX=auxiliary, like "is""has""will""should";
## (continued): PRON=prounoun; SCONJ=subordinating conjunction, like "if""while""that"; ADV=adverb;
## (continued): INTJ=interjection, like "psst""ouch""bravo""hello"; PUNCT=punctuation; SYN=symbol; X = other

## Also notice that, even if I have removed all punctuation in the text preparation step, 
## punctuation still makes up 3.49% of the text. It might happen because the punctuation library 
## does not contain certain variants/forms of the punctuation in the text. This situation is worthy to notice
## as more text cleaning processes can still be done, but it should not affect the final result a lot as we 
## are recording percetages here. 


The above process will be repeated for the whole text and the result is mannually recorded and updated in a Microsoft Excel document. See "Parsing Result.xlsx".

For the length of this notebook, these steps are done in another notebook named "Auxilaiary Text Processing."

### Parsing Result

The "ParsingResult.xlsx" is saved as a .csv file for better used in jupyter notebook.

In [32]:
## import the .csv file
parse = pd.read_csv("ParsingResult.csv")
print(parse)

   Unnamed: 0          A          B          C          D          E  \
0        NOUN      14.25      11.49      11.75      12.32      12.23   
1       PROPN       5.59       5.71       5.42       5.21       7.20   
2        VERB      13.87      13.94      13.81      14.96      14.98   
3         DET       9.15       7.37       7.57       8.20       8.07   
4        PART       3.30       3.19       3.29       3.80       3.55   
5         ADP      10.87       9.05       9.39      10.24       9.67   
6       CCONJ       3.46       2.88       2.90       3.12       3.47   
7         AUX       4.81       4.87       4.99       5.47       5.09   
8         ADJ       5.28       4.44       4.52       4.75       4.80   
9        PRON      10.54      11.52      11.31      12.43      11.50   
10      SCONJ       1.47       1.44       1.64       1.77       1.30   
11        ADV       5.79       5.59       5.74       6.15       5.98   
12       INTJ       0.18       0.24       0.20       0.21       

Data from all 6 parts are combined to yield a final percentage breakdown. Calculation is all done in the Excel file.

In [31]:
result_combined = pd.read_csv("PercentageResult.csv")
print(result_combined)


   Combined percentage            A+B            C+D            E+F
0                 NOUN      12.799715      12.041632      12.415217
1                PROPN       5.653056       5.312557       7.014783
2                 VERB      13.906783      14.398379      14.856522
3                  DET       8.214671       7.892330       8.023696
4                 PART       3.242199       3.550933       3.539710
5                  ADP       9.913652       9.824889       9.618551
6                CCONJ   18583.823910       3.012560       3.629492
7                  AUX       4.841528       5.235584       5.239203
8                  ADJ       4.838609       4.637676       4.887464
9                 PRON      11.054956      11.883030      11.546304
10               SCONJ       1.454236       1.706512       1.356594
11                 ADV       5.684907       5.949770       6.077753
12                INTJ       0.211528       0.205116       0.238841
13               Total  270933.000000  321636.00

# Data Analysis

In Chan's analysis, he pays attention to four categories of POS: adjective, noun, adverb and stative verb.

#### Adjective

My result shows the occurrence frequency of adjective is 4.84%, 4.64% and 4.89% (the 8th row). They are very close to each other, and it is hard to conclude that E+F(the last fourty chapters) is significantly different from A+B and C+D. 

#### Noun

The occurrence frequency of noun is 12.80%, 12.04% and 12.42%. They are very close to each other, and it is hard to conclude that E+F(the last fourty chapters) is significantly different from A+B and C+D.

We should also notice that in English, pronoun makes up a significant part in POS(5.65%, 5.31% and 7.01%). In modern Chinese, however, pronoun is not distinguishable from normal noun; in classical Chinese, some words are more likely to be used in the position of pronoun and maybes the equivalence of noun. (For the pronoun used in Classical Chinese, see Michael A. Fuller's *An Introduction to Literary Chinese*, page 262.) In Chan's research, he defines a noun as "a word used for naming some person or thing" (Chan 80). Therefore, pronoun is overlooked when he was counting. Overall, I will omit pronoun when considering the occurance frequency of pronoun.

#### Adverb

The occurrence frequency of adverb is 5.68%, 5.95% and 6.08%. They are very close to each other, and it is hard to conclude that E+F(the last fourty chapters) is significantly different from A+B and C+D.

#### Stative Verb

"A stative verb is one that describes a state of being, in contrast to a dynamic verb, which describes an action. The difference can be categorized by saying that stative verbs describe situations that are static or unchanging throughout their entire duration, whereas dynamic verbs describe processes that entail change over time" (Wikipedia https://en.wikipedia.org/wiki/Stative_verb). 
In English, stative verb is made up of a broad type of verbs. For examples of stative verb in English, see: https://www.thoughtco.com/stative-verb-1692139.

Chan's paper defines words that qualify the following rules as stative verb (Chan 93)
1. A stative verb can be preceded by a word meaning "very"
2. It does not take any object.
3. It describes a state of being and thus in meaning resemble an adjective in the predicative use in an English sentence.
Thus, stative verb in Chinese combines words that are from both the adjetive and verb categories. It is hard for us to use the result of the POS analysis to give the occurrence frequency of stative verb.

# Conclusion

From the above analysis, we can conclude that the three parts - A+B, C+D and E+F - are not significantly different from one another, indicating the "likeness/closeness between the vocabularies of texts" (Chan 117). As we clearly know that the English translation is done by two translators, why does the computational test fail to recognize the difference and give “the wrong result”?

Unlike Chan's result, which reveals that the extinct 120-chapter version is by the same author, my result can be interpreted in multiple ways. In the following section, I provide three hypotheses for explaining why I obtain this result. 
By providing possible explanations, I do not intend to prioritize one over the other. Instead, the purpose of listing these hypotheses is to evoke more meaningful questions about the computational analysis of style and the question of authorship.

### Hypothesis 1: Mastery of Translation

In this hypothesis, I assume that Chan’s method is performed perfectively and effectively gives an accurate result on authorship. 

As mentioned earlier, David Hawkes and John Minford work collaboratively on the translation. Minford actively participated in the translation of the first eighty chapters, and he translated the last forty chapters not too long after Hawkes completed the first eighty chapters. Is it possible that the two translators’ style is so similar that it can even pass the computational analysis? 

Accepting this hypothesis immediately questions the validity of Chan’s conclusion. If it is possible for Minford to translate in Hawkes’ style, could someone successfully imitate Cao Xueqin’s style and write the last forty chapters and pass the computational test? Therefore, the question of authorship seems to be proved in vain. Why should we care about authorship if the “another author” can perfectively write in the style of “the author”?

### Hypothesis 2: Problems in Chan’s Research

In this hypothesis, I assume the “wrong result” occurs due to the problems in Chan’s method. By repeating the wrong method, we can arrive only at the wrong result. 

Chan’s research assumes that different writers have their distinct styles, and the differences can be expressed as the way their sentences are constructed and therefore, reflected in the occurrence frequency of certain POS. Theoretically, it is impossible to deceive the POS test because no one write while calculating the POS makeup of their sentences, making POS test the best parameter for tagging an author’s writing style. 

Chan’s result, however, is challenged by Chao Kang’s research, who analyzed a different POS, the empty word (xuci). Empty words in Classical Chinese do not carry any semantic meaning but functions as syntactic markers. Because of the lack of sematic meaning, in many situations, empty words can be omitted according to the author’s will. Therefore, Chao considers empty words the best indicator of an author’s style, and by analyzing empty words, Chao ends by concluding the presence of different authors. 

Chan’s and Chao’s conclusions are equally compelling, as they followed a very similar method, that is, to statistically analyzes the characteristic of the text’s POS, but chose different POS for analyzing. This gives rise to the question that whose POS is better and how to determine a better POS.


### Hypothesis 3: Problems in My Research

In this hypothesis, I assume Chan’s method gives correct results, but my process fails to repeat his. 
Chan’s research is executed in more detail than mine. For instance, Chan takes account of “noun-condensed” word so that these words will only be counted once. The word “problem child,” for instance, is counted as two nouns by spaCy although it should be only counted once. 

Moreover, Chan performed more statistical analysis to test the statistical validity. In my research, I only measure the frequency of total occurrence of certain POS. In contrast, Chan collected the relative frequency of words and performed association test to testify the likeliness. (I attempted to repeat Chan’s method but was not able to continue due to spaCy’s limitation of text length. Additional work can be done by further breaking down the text into shorter parts.)  
